### PR TITLE
Added Exit Confirmation Popup

### PR DESCRIPTION
**Pull request by rsashank**
### What does this PR do, and why?
Displays a confirmation popup when the (^C) interrupt signal is encountered, seeking confirmation prior to exiting the terminal client.

### Outstanding aspect(s)    

- [ ] 

### External discussion & connections

- [ ] Discussed in **#zulip-terminal** in `topic`
- [ ] Fully fixes #
- [x] Partially fixes issue #1341
- [ ] Builds upon previous unmerged work in PR #
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### How did you test this?

- [x] Manually - Behavioral changes
- [x] Manually - Visual changes
- [x] Adapting existing automated tests
- [ ] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [x] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [ ] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [ ] It has a commit summary describing the  motivation and reasoning for the change
- [x] It individually passes linting and tests
- [ ] It contains test additions for any new behavior
- [ ] It flows clearly from a previous branch commit, and/or prepares for the next commit

### Visual changes    

https://github.com/zulip/zulip-terminal/assets/110327345/f95a3f00-88ec-40cb-91ae-5ec8890a5985

### CODE REVIEW

File: zulipterminal/core.py [hunk 1]
```diff
@@ -617,11 +617,18 @@ def narrow_to_all_mentions(self) -> None:
         self._narrow_to(anchor=None, mentioned=True)
 
     def deregister_client(self) -> None:
-        queue_id = self.model.queue_id
-        self.client.deregister(queue_id, 1.0)
+        raise urwid.ExitMainLoop
```
Comment (neiljp) [inline zulipterminal/core.py:620]: This does seem cleaner than the sys.exit we currently use.

However, without that I've experienced strange states where the client doesn't exit properly.

File: zulipterminal/core.py [hunk 2]
```diff
@@ -617,11 +617,18 @@ def narrow_to_all_mentions(self) -> None:
...
+        question = urwid.Text(
+            ("bold", "Do you want to deregister the client?"),
+            "center",
+        )
+        popup_view = PopUpConfirmationView(
+            self, question, self.deregister_client, location="center"
+        )
+        self.loop.widget = popup_view
+        self.loop.run()
         sys.exit(0)
```
Comment (neiljp) [inline zulipterminal/core.py:632]: I was surprised that we need the `loop.run()` here, but perhaps the signal handling interrupts the loop in some way. It seems to behave strangely without it?

**Review by neiljp — COMMENTED** [review]
@rsashank Thanks for posting the code :+1:

This does appear to work quite well, though note that this has a regression in that it doesn't let the server know that we no longer need the queue. You can see this if you run in debug mode (-d) and look at the generated API log file.

There seem to be some nuances in using the signal handler with urwid code in the event loop. I want to like using `ExitMainLoop` alone for exiting, but it doesn't appear to reliably clear up properly in the absence of the existing `sys.exit`.

File: zulipterminal/core.py (hunk 2 above)
Comment (rsashank) [reply zulipterminal/core.py:634]: Thank you for your comment and for reviewing the code. I appreciate your feedback.
The absence of `loop.run()` results in the popup not appearing, it appears that you are correct, the signal handling is interrupting the loop in some way.

File: zulipterminal/core.py (hunk 1 above)
Comment (rsashank) [reply zulipterminal/core.py]: Thank you for your comment and for reviewing the code. I appreciate your feedback.
I've reverted to the existing code, the behavior seems apt now.

**Comment by rsashank** [comment]
I've reverted to the original code for the `deregister_client` method. I believe it now results exits the client cleanly.

**Comment by neiljp** [comment]
@rsashank Thanks for working on this and the update :+1:

Did you test this? The version you pushed had an explicit call to deregister_client from the signal handler before switching to the confirmation popup, so always exits before it has a chance to enter the popup.

I just pushed a version here which removes that line, as well as the existing sys.exit in the exit_handler, which should now be unnecessary. This version also has a user-friendly exit message - I would expect most users will not know about deregistering vs simply exiting :)

With an updated commit message, I plan to merge this shortly.

I was able to trigger a crash in an edge case of overusing ^C, which I suspect is related to the UI in the signal handler, but this works for now. We'll want to ensure we can configure this behavior next, as per the issue, and possibly discuss whether the default should be yes or no.

**Comment by rsashank** [comment]
@neiljp Thanks for reviewing the code. 

I was testing calling the method in the exit_handler and inadvertently force-pushed the test code. I did test the code prior to that. I apologize for the novice oversight.

 I'll look into the edge case of overusing ^C, and will be looking forward to working on configuring this behavior next.

Thanks for the guidance, this is my first contribution to open source. Looking forward to the merge. :)

**Comment by neiljp** [comment]
@rsashank Thanks for the contribution - I know it feels good to have code merged :)

As per my post in the stream, I suspect this will now be important to be configurable, initially in the config file, whether for existing users or developers who know what they are doing. Hopefully comments in the stream will help guide the importance of that.

You're welcome to dig into the ^C issue, but it seemed quite challenging to trigger, so I wouldn't think of it as a priority in comparison.
